include/mbgl/gfx/drawable.hpp

```diff
@@ -0,0 +1,230 @@
+#pragma once
+
+#include <mbgl/gfx/uniform_buffer.hpp>
+#include <mbgl/tile/tile_id.hpp>
+#include <mbgl/util/color.hpp>
+#include <mbgl/util/identity.hpp>
+#include <mbgl/util/traits.hpp>
+#include <mbgl/gfx/texture2d.hpp>
+#include <mbgl/gfx/drawable_data.hpp>
+
+#include <cstdint>
+#include <memory>
+#include <optional>
+#include <vector>
+
+namespace mbgl {
+
+class Color;
+class PaintParameters;
+enum class RenderPass : uint8_t;
+
+namespace gfx {
+
+class ColorMode;
+class CullFaceMode;
+enum class DepthMaskType : bool;
+class DrawableTweaker;
+class DrawMode;
+class IndexVectorBase;
+class ShaderProgramBase;
+class VertexAttributeArray;
+
+using DrawPriority = int64_t;
+using DrawableTweakerPtr = std::shared_ptr<DrawableTweaker>;
+using IndexVectorBasePtr = std::shared_ptr<IndexVectorBase>;
+using ShaderProgramBasePtr = std::shared_ptr<ShaderProgramBase>;
+using Texture2DPtr = std::shared_ptr<Texture2D>;
+
+class Drawable {
```

Comment:
Some of these are things that can be modified on the fly.  So yes to the tweakers question.
Some is just convenience, I think.  Easier to build the Drawable up in pieces.

Reply:
So Drawable Builders are kind of the step by step object you use to build drawables.  At least that's the intent.  The way Buckets worked made that a bit more complicated, but there will be more on that in the Layer PRs.

I can see some of the set methods you would never expect to change after creation (isCustom) and some you would.

Now I don't think we need to get strict with this (ala constructors) but it would be helpful to mark which ones are expected to change and which ones aren't.  @TimSylvester  or @mwilsnd thoughts on this?

Reply:
The one pattern I find myself gravitating towards in cases like this is a configuration struct given to an object's constructor. Vulkan leverages this pattern extensively, naming such objects `<x>CreateInfo`. In essence, a relevant CreateInfo struct is populated and then passed (or moved) to the constructor of an object, passing all the data which should be immutable (or at least private) for the object's lifetime. Any remaining accessor methods for things not covered by a creation struct are for truly dynamic state.

Reply:
That is a nice approach for a rendering SDK.
I don't really want to restructure the Drawable around that idea, though.  It seems like a fair bit of work for a very low level utility object here.

Reply:
The Circle Layer is one of the simpler implementations as far as layers go:

https://github.com/maplibre/maplibre-native/blob/topic/drawable/src/mbgl/renderer/layers/render_circle_layer.cpp
Look around 395 in the update call.  You'll see that we don't really build drawables directly, we use the Builders.